hugolib/site_sections.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2017 The Hugo Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hugolib
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+
+	radix "github.com/hashicorp/go-immutable-radix"
+	"github.com/spf13/hugo/helpers"
+)
+
+// Deprecated: Use .Site.Home.Sections.
+// TODO(bep) Hugo 0.23 = Reuse as an alias for home's sections.
+func (s *SiteInfo) Sections() Taxonomy {
+
+	helpText := `In Hugo 0.22 we introduced nested sections, so this method now returns an empty taxonomy.
+	
+To list sections with its pages, you can do something like this:
+
+{{ range $.Site.Home.Sections }}
+  Section: {{ .Title }}
+  {{ range .Pages }}
+     Section Page: {{ .Title }}
+  {{ end }}
+{{ end }}
+
+To get a specific section, you can do this:
+
+{{ $section := $.Site.GetPage "section" "blog" }}
+`
+
+	helpers.Deprecated("Site", "Sections", helpText, true)
+
+	return Taxonomy{}
+}
+
+// Home is a shortcut to the home page, equivalent to .Site.GetPage "home".
+func (s *SiteInfo) Home() (*Page, error) {
+	return s.GetPage(KindHome)
+}
+
+// Parent returns a section's parent section or a page's section.
+// To get a section's subsections, see Page's Sections method.
+func (p *Page) Parent() *Page {
+	if p.Kind == KindPage {
+		return p.currentSection
+	}
+	return p.parent
+}
+
+// current returns the page's current section.
+// Note that this will return nil for pages that is not regular, home or section pages.
+// Note that for paginated sections and home pages, this will return the original page pointer.
+func (p *Page) current() *Page {
+	v := p
+	if v.origOnCopy != nil {
+		v = v.origOnCopy
+	}
+	if v.IsHome() || v.IsSection() {
+		return v
+	}
+
+	return v.currentSection
+}
+
+// InSection returns whether the given page is in the current section.
+// Note that this will always return false for pages that are
+// not either regular, home or section pages.
+func (p *Page) InSection(other interface{}) (bool, error) {
+	if p == nil || other == nil {
+		return false, nil
+	}
+
+	if po, ok := other.(*PageOutput); ok {
+		other = po.Page
+
+	}
+
+	pp, ok := other.(*Page)
+	if !ok {
+		return false, fmt.Errorf("%T not supported in IsActive", other)
+	}
+
+	if pp == nil {
```

Comment:
Go question:  is it possible for pp to be nil here?  Already checked `other != nil` and `!ok`.